sw/device/lib/crypto/aes_gcm/aes_gcm.c

```diff
@@ -0,0 +1,351 @@
+ * field element is shifted, it can be reduced by multiplying the high bits
+ * (128 and above) by 0xe1 and adding them to the low bits.
+ *
+ * There is a size/speed tradeoff in window size. For 8-bit windows, GHASH
+ * becomes significantly faster, but the overhead for computing the product
+ * table of a given hash subkey becomes higher, so larger windows are slower
+ * for smaller inputs but faster for large inputs.
+ */
+static const uint16_t kGFReduceTable[16] = {
+    0x0000, 0x201c, 0x4038, 0x6024, 0x8070, 0xa06c, 0xc048, 0xe054,
+    0x00e1, 0x20fd, 0x40d9, 0x60c5, 0x8091, 0xa08d, 0xc0a9, 0xe0b5};
+
+/**
+ * Performs a bitwise XOR of two blocks.
+ *
+ * This operation corresponds to addition in the Galois field.
+ *
+ * @param condition Either 0 or 1
+ * @param x First operand block
+ * @param y Second operand block
+ * @param out Buffer in which to store output; can be the same as one or both
+ * operands.
+ */
+static inline void block_xor(const aes_block_t *x, const aes_block_t *y,
+                             aes_block_t *out) {
+  for (size_t i = 0; i < kAesBlockNumWords; i++) {
+    out->data[i] = x->data[i] ^ y->data[i];
+  }
+}
+
+/**
+ * Reverse the bits of a 4-bit number.
+ *
+ * @param word Input word.
+ * @return Word with bytes reversed.
+ */
+static uint8_t reverse_bits(uint8_t byte) {
+  uint8_t out = 0;
+  for (size_t i = 0; i < 4; i++) {
+    out <<= 1;
+    out |= (byte >> i) & 1;
+  }
+  return out;
+}
+
+/**
+ * Reverse the bytes of a 32-bit word.
+ *
+ * This function is convenient for converting between the processor's native
+ * integers, which are little-endian, and NIST's integer representation, which
+ * is big-endian.
+ *
+ * @param word Input word.
+ * @return Word with bytes reversed.
+ */
+static uint32_t reverse_bytes(uint32_t word) {
+  uint32_t out = 0;
+  for (size_t i = 0; i < sizeof(uint32_t); i++) {
+    out <<= 8;
+    out |= (word >> (i << 3)) & 255;
+  }
+  return out;
+}
+
+/**
+ * Logical right shift of an AES block.
+ *
+ * NIST represents AES blocks as big-endian, and a "right shift" in that
+ * representation means shifting out the LSB. However, the processor is
+ * little-endian, so we need to reverse the bytes before shifting.
+ *
+ * @param block Input AES block, modified in-place
+ * @param nbits Number of bits to shift
+ */
+static inline void block_shiftr(aes_block_t *block, size_t nbits) {
+  // To perform a "right shift" with respect to the big-endian block
+  // representation, we traverse the words of the block and, for each word,
+  // reverse the bytes, save the new LSB as `carry`, shift right, set the MSB
+  // to the last block's carry, and then reverse the bytes back.
+  uint32_t carry = 0;
+  for (size_t i = 0; i < kAesBlockNumWords; i++) {
+    uint32_t rev = reverse_bytes(block->data[i]);
+    uint32_t next_carry = rev & ((1 << nbits) - 1);
+    rev >>= nbits;
+    rev |= carry << (32 - nbits);
+    carry = next_carry;
+    block->data[i] = reverse_bytes(rev);
+  }
```

Comment:
nit: can't we do this in the opposite order? This would eliminate two calls to `reverse_bytes()`.

Reply:
Yes, but it's much less readable. Now that I have a working version I'll give it another shot and see if it makes enough speed difference to be worthwhile!